lib/puppet/configurer.rb

```diff
@@ -222,8 +222,7 @@ def run(options = {})
           found = find_functional_server()
           server = found[:server]
           if server.nil?
-            Puppet.warning _("Could not select a functional puppet master")
-            server = [nil, nil]
+            raise Puppet::Error, _("Could not select a functional puppet master from server_list")
```

Comment:
How about appending `Puppet.settings[:server_list]` to the error message, so we know what servers it tried?

Reply:
Done